### ISSUE TITLE
export onnx have some bug

### ISSUE THREAD

**Issue by WangHHY19931001** [ultralytics/ultralytics#23055]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

yolo12n export onnx used by android onnxruntime with nnapi ep get error:
``` shell
libc++abi: terminating due to uncaught exception of type Ort::Exception: op_builder_helpers.cc:144 AddNnapiSplit count [0] does not evenly divide dimension 3 [32]
```
may be export has some error

### Environment

Ultralytics 8.3.241  Python-3.12.12 torch-2.9.1+cpu CPU (12th Gen Intel Core(TM) i5-12400)
YOLOv12n summary (fused): 159 layers, 2,563,163 parameters, 0 gradients, 6.4 GFLOPs

[34m[1mPyTorch:[0m starting from 'best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 300, 6) (15.4 MB)

[34m[1mONNX:[0m starting export with onnx 1.20.0 opset 22...
C:\Users\dell\miniconda3\envs\yolo\Lib\site-packages\torch\onnx\_internal\torchscript_exporter\utils.py:1447: OnnxExporterWarning: Exporting to ONNX opset version 22 is not supported. by 'torch.onnx.export()'. The highest opset version supported is 20. To use a newer opset version, consider 'torch.onnx.export(..., dynamo=True)'.
  warnings.warn(
[34m[1mONNX:[0m export success  1.1s, saved as 'best.onnx' (10.1 MB)

Export complete (1.3s)
Results saved to [1mD:\ultralytics[0m
Predict:         yolo predict task=detect model=best.onnx imgsz=640
Validate:        yolo val task=detect model=best.onnx imgsz=640 data=/mnt/d/yolo/ultralytics/datasets/YOLODataset/dataset.yaml
Visualize:       https://netron.app

### Minimal Reproducible Example

export python:
``` python
import os.path

from ultralytics import YOLO
import onnx
import numpy as np
import torch

model = YOLO("best.pt")  # load a custom trained model
model.export(
    format="onnx",      # 导出格式为 torchscript
    imgsz=(640, 640),   # 设置输入图像的尺寸
    keras=False,        # 不导出为 Keras 格式
    optimize=False,     # 不进行优化 False, 移动设备优化的参数，用于在导出为TorchScript 格式时进行模型优化
    half=False,          # 启用 FP16 量化
    int8=False,         # 不启用 INT8 量化
    dynamic=False,      # 不启用动态输入尺寸
    simplify=False,      # 简化 ONNX 模型
    opset=None,         # 使用最新的 opset 版本
    nms=True,          # 不添加 NMS（非极大值抑制）
    batch=1,            # 指定批处理大小
    device="cpu"        # 指定导出设备为CPU或GPU，对应参数为"cpu" , "0"
)
```

use c++ code:
```
    env = new Ort::Env(ORT_LOGGING_LEVEL_WARNING, "Default");
    Ort::SessionOptions so;
    so.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);

    uint32_t nnapi_flags = 0;
    nnapi_flags |= NNAPI_FLAG_USE_FP16;
    Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_Nnapi(so, nnapi_flags));
    LOGI("OttCheck set onnxruntime session opt end");
    session =  new Ort::Session(*env, model_data.data(), model_data.size(), so);
    LOGI("OttCheck create onnxruntime session (load model)");
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @WangHHY19931001, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.  

This is an automated response 🤖—an Ultralytics engineer will also assist you soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. You’ve included a solid start already—thank you! 🙏 To make it fully reproducible across environments, please also add:
-Your exact export command (CLI) or a complete Python script (including the `ultralytics` version printout)
-The exported `best.onnx` (or a small public model + steps to reproduce)
-Your Android onnxruntime version + device model + Android version
-Your exact NNAPI settings and whether CPU EP works with the same ONNX
-The full stack trace/log around `AddNnapiSplit ... does not evenly divide dimension ...`

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can export without `nms=True`

**Comment by WangHHY19931001** [comment]
> You can export without `nms=True`

i try it but still error, got error
``` shell
libc++abi: terminating due to uncaught exception of type Ort::Exception: op_builder_helpers.cc:144 AddNnapiSplit count [0] does not evenly divide dimension 3 [32]
```

**Comment by Y-T-G** [comment]
Set `simplify=True`

**Comment by WangHHY19931001** [comment]
> Set `simplify=True`

i try it but still got error
``` shell
libc++abi: terminating due to uncaught exception of type Ort::Exception: op_builder_helpers.cc:144 AddNnapiSplit count [0] does not evenly divide dimension 3 [32]
```
may be we should to check modle?

**Comment by glenn-jocher** [comment]
That NNAPI EP crash looks like a provider-compile limitation, and your export is also using opset 22 even though PyTorch warns the classic exporter only supports up to opset 20, so please re-export with an explicit opset and no NMS (see the [Export mode docs](https://docs.ultralytics.com/modes/export/) and [`Model.export`](https://docs.ultralytics.com/reference/engine/model/) signature):  

```python
from ultralytics import YOLO

model = YOLO("best.pt")
model.export(format="onnx", opset=20, nms=False, simplify=True, dynamic=False, device="cpu")
```

If NNAPI still fails after `opset=20` (or try `opset=17`), it’s likely YOLO12-specific ops aren’t supported by NNAPI; please confirm whether the same ONNX runs with the default CPU EP on Android, and paste `yolo checks` plus your Android onnxruntime version and full text logs so we can reproduce accurately.

**Comment by WangHHY19931001** [comment]
> That NNAPI EP crash looks like a provider-compile limitation, and your export is also using opset 22 even though PyTorch warns the classic exporter only supports up to opset 20, so please re-export with an explicit opset and no NMS (see the [Export mode docs](https://docs.ultralytics.com/modes/export/) and [`Model.export`](https://docs.ultralytics.com/reference/engine/model/) signature):
> 
> from ultralytics import YOLO
> 
> model = YOLO("best.pt")
> model.export(format="onnx", opset=20, nms=False, simplify=True, dynamic=False, device="cpu")
> If NNAPI still fails after `opset=20` (or try `opset=17`), it’s likely YOLO12-specific ops aren’t supported by NNAPI; please confirm whether the same ONNX runs with the default CPU EP on Android, and paste `yolo checks` plus your Android onnxruntime version and full text logs so we can reproduce accurately.

use export python code:
``` python
import os.path

from ultralytics import YOLO
import onnx
import numpy as np
import torch

model = YOLO("best.pt")  # load a custom trained model
model.export(
    format="onnx",      # 导出格式为 torchscript
    imgsz=(640, 640),   # 设置输入图像的尺寸
    keras=False,        # 不导出为 Keras 格式
    optimize=False,     # 不进行优化 False, 移动设备优化的参数，用于在导出为TorchScript 格式时进行模型优化
    half=False,          # 启用 FP16 量化
    int8=False,         # 不启用 INT8 量化
    dynamic=False,      # 不启用动态输入尺寸
    simplify=False,      # 简化 ONNX 模型
    opset=None,         # 使用最新的 opset 版本
    nms=True,          # 不添加 NMS（非极大值抑制）
    batch=1,            # 指定批处理大小
    device="cpu"        # 指定导出设备为CPU或GPU，对应参数为"cpu" , "0"
)
```

android onnxruntime cpu ep run it success

android onnxruntime nnapi ep run it get error

onnxruntime version:
a83fc4d58cb48eb68890dd689f94f28288cf2278

build cmd:
``` cmd
.\build.bat --config Release --android --android_sdk_path "C:\Users\dingli\AppData\Local\Android\Sdk" --android_ndk_path "C:\Users\dingli\AppData\Local\Android\Sdk\ndk\27.3.13750724" --android_abi arm64-v8a --android_api 29 --cmake_generator Ninja --use_nnapi --skip_tests --cmake_extra_defines "CMAKE_CXX_FLAGS=-flto -ffast-math -funroll-loops -march=armv8.2-a+fp16 -falign-loops=16 -falign-functions=16 -ftree-vectorize -mllvm -polly" "CMAKE_C_FLAGS=-flto -ffast-math -funroll-loops -march=armv8.2-a+fp16 -falign-loops=16 -falign-functions=16 -ftree-vectorize -mllvm -polly" "CMAKE_INTERPROCEDURAL_OPTIMIZATION=ON"
```

that all i try

**Comment by WangHHY19931001** [comment]
``` shell
yolo checks
Ultralytics 8.3.241  Python-3.12.12 torch-2.9.1+cpu CPU (12th Gen Intel Core(TM) i5-12400)
Setup complete  (12 CPUs, 31.7 GB RAM, 32.1/50.0 GB disk)

OS                     Windows-10-10.0.19045-SP0
Environment            Windows
Python                 3.12.12
Install                pip
Path                   C:\Users\dell\miniconda3\envs\yolo\Lib\site-packages\ultralytics
RAM                    31.70 GB
Disk                   32.1/50.0 GB
CPU                    12th Gen Intel Core(TM) i5-12400
CPU count              12
GPU                    None
GPU count              None
CUDA                   None

numpy                   2.2.6>=1.23.0
matplotlib              3.10.8>=3.3.0
opencv-python           4.12.0.88>=4.6.0
pillow                  12.0.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.16.3>=1.4.1
torch                   2.9.1>=1.8.0
torch                   2.9.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.24.1>=0.9.0
psutil                  7.2.0>=5.8.0
polars                  1.36.1>=0.20.0
ultralytics-thop        2.0.18>=2.0.18
```

use opset 20 with nnapi still got same error:
``` shell
libc++abi: terminating due to uncaught exception of type Ort::Exception: op_builder_helpers.cc:144 AddNnapiSplit count [0] does not evenly divide dimension 3 [32]
```

use opset 17 with nnapi success

**Comment by WangHHY19931001** [comment]
thank's @Y-T-G  and @glenn-jocher 

bug fund, now use opset 17 export can fix it. please close this issue and add some helpful into export onnx

**Comment by glenn-jocher** [comment]
Good catch — this looks like an Android NNAPI EP limitation with newer ONNX opsets for this graph, so using `opset=17` is the right workaround (higher opsets may still run fine on the CPU EP); we’ll add a note to the [Export mode docs](https://docs.ultralytics.com/modes/export/) and the [`best_onnx_opset()` reference](https://docs.ultralytics.com/reference/engine/exporter/) for clarity, and we can close this as resolved.
